### PR TITLE
remove single-view assumption from `paintImage`

### LINKED ISSUE #117477: Remove hard-coded devicePixelRatio from `paintImage`
`paintImage` currently hard-codes the device pixel ratio by reaching out to the window singleton:

https://github.com/flutter/flutter/blob/d1244b7c9b8d676a84bdba2a440ddae1e0907a8c/packages/flutter/lib/src/painting/decoration_image.dart#L565-L567

In a multi-viewed world, this will not work anymore because `paintImage` has no idea what view the image is getting painted on. This will need to be refactored to get the devicePixelRatio passed in.

### PR DESCRIPTION
**Pull request by goderbauer**
Fixes https://github.com/flutter/flutter/issues/117477.
Part of https://github.com/flutter/flutter/issues/116929.

### CODE REVIEW

**Comment by goderbauer** [comment]
@dnfield A while ago you added this code to `paintImage` that checks for oversized images. This debug check is the only reason, why the `paintImage` method needs to know the device pixel ratio, which it is currently reading from the global window singleton. That will not work in a multi window/view world. The method needs to read the dpr from the view the image is going to be drawn into. Unfortunately, it doesn't have access to that information. In this PR I changed it so that the dpr can (optionally) be passed into the method to enable the debug check and I made sure that everywhere in the framework we do pass it in. However, if somebody now calls `paintImage` directly without passing in the DPR, the oversized images check is disabled.

Another option would be to make passing in the dpr required. This would be a breaking change since there are two callers of this method in google3 (those could be easily updated, though). It does feel a little odd to require it just for the debug check, though.

Yet another option could be to move the check out of paintImage. However, I didn't really find a better place for it...

What are your thoughts?

File: packages/flutter/lib/src/painting/decoration_image.dart [hunk 1]
```diff
@@ -424,6 +424,10 @@ void debugFlushLastFrameImageSizeInfo() {
 ///
 ///  * `image`: The image to paint onto the canvas.
 ///
+///  * `devicePixelRatio`: The device pixel ratio of the [FlutterView] into
+///     which the image will be drawn. Can be obtained via
+///     `View.of(context).devicePixelRation`.
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/decoration_image.dart:429]: note to self: mention `MediaQuery.devicePixelRatioOf(context)` as well

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -5980,6 +5981,7 @@ class RawImage extends LeafRenderObjectWidget {
       invertColors: invertColors,
       isAntiAlias: isAntiAlias,
       filterQuality: filterQuality,
+      devicePixelRatio: View.maybeOf(context)?.devicePixelRatio,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:5984]: note to self: here and below use `MediaQuery.maybeDevicePixelRatioOf(context)` instead.

**Comment by goderbauer** [comment]
"Google testing" failed because tests were run without including prerequisite https://github.com/flutter/flutter/pull/117244.

File: packages/flutter/lib/src/painting/decoration_image.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/decoration_image.dart:429]:
```suggestion
///     `View.of(context).devicePixelRatio`.
```

Comment (dnfield) [reply packages/flutter/lib/src/painting/decoration_image.dart:429]: Can this get defaulted somehow?

For example, just default it to the first view in the collection of views?

Comment (goderbauer) [reply packages/flutter/lib/src/painting/decoration_image.dart:429]: If we default this, the assert check in `paintImage` becomes more or less meaningless, no? It needs to know the exact dpr in order to figure out whether the current image dimensions are overkill or not.

Comment (goderbauer) [reply packages/flutter/lib/src/painting/decoration_image.dart:429]: We could just use the max device pixel ratio of all views as an upper bound, though. That will greatly simplify things.

**Comment by goderbauer** [comment]
Superseded by https://github.com/flutter/flutter/pull/118721.

**Pull request closed without merging**